### PR TITLE
Built-in deploy generator

### LINKED ISSUE #4847: Deployment generators to export package contents to a specific layout
Coming from #3243 and related to #4827 and the default deployable concept in #3280.

I have played witht the idea of having a generator that deploys the files of a dependency tree to a specific layout. They idea is, instead of using the paths to generate a file, use them to copy the needed files to the installation folder.

In order to achive this you would need to create a "dumb recipe" (*conanfile.txt) with the requirements of the package you want to deploy ("ncurses" in this case)
and the custom generator (called "deployer" in this case).

*conanfile.txt*
```
[requires]
ncurses/6.1@danimtb/testing

deployer/0.1@danimtb/testing

[generators]
deployer
```

The custom generator I created as a example is the following:

*conanfile.py*
```
import os
import shutil

from conans.model import Generator
from conans import ConanFile


class deployer(Generator):
    @property
    def filename(self):
        return "conandeployer.txt"

    @property
    def content(self):
        for dep_name, dep_cpp_info in self.deps_build_info.dependencies:
            if dep_name == "deployer":
                continue
            for key in ["include_paths", "lib_paths", "bin_paths"]:
                if getattr(dep_cpp_info, key):
                    for path in getattr(dep_cpp_info, key):
                        if os.listdir(path):
                            src = os.path.normpath(path)
                            dir_name = os.path.basename(path)
                            dst = os.path.join(self.output_path, dep_name, dir_name)
                            shutil.copytree(src, dst)
        return "Files deployed!"


class MyCustomGeneratorPackage(ConanFile):
    name = "deployer"
    version = "0.1"
```
There are some tricks to make the generator work, for example using the ``self.output_path`` attribute.

However, this approach can be applied externally without modifying the recipe of the package to be deployed.

Ideas that this issue tries to merge:
- "Deploy packages to a custom structure (like *chroot*)" by @brianthelion in #3243 
- "Deploy packages to a custom structure using hooks" by @brianthelion in #4827
- "Having a default deploy method in recipes" by @solvingj in  #3280
- "Using a proxy package with a deploy method" by @monsdar in  #3280 (this idea is pretty similar to the one exposed here).

### PR DESCRIPTION
**Pull request by danimtb**
Changelog: Feature: New ``deploy`` generator to export files from a dependency graph to an installation folder
Docs: https://github.com/conan-io/docs/pull/1262

- [x] Refer to the issue that supports this Pull Request: closes #4847
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [x] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

### CODE REVIEW

File: conans/client/generators/deploy.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    @property
+    def filename(self):
+        return BUILD_INFO_DEPLOY
+
+    @property
+    def content(self):
+        copied_files = []
+
+        for dep_name, dep_cpp_info in self.deps_build_info.dependencies:
+            for directory_path in self.directory_path_names:
```
Comment (lasote) [inline conans/client/generators/deploy.py]: Why this if we are already doing the bulk copy?

File: conans/client/generators/deploy.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        date = calendar.timegm(time.gmtime())
+        file_dict = {}
+        for f in copied_files:
+            abs_path = os.path.join(self.output_path, f)
+            file_dict[f] = md5sum(abs_path)
+        manifest = FileTreeManifest(date, file_dict)
+        return repr(manifest)
+
+    @property
+    def directory_path_names(self):
```
Comment (lasote) [inline conans/client/generators/deploy.py]: Why this if we are already doing the bulk copy?

File: conans/test/functional/generators/deploy_test.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+            self.assertNotIn(name, content)
+
+
+class DeployGeneratorGraphTest(unittest.TestCase):
+    """
+    Deploy generator set of tests with more than one package in the graph
+    """
+
+    def setUp(self):
+        conanfile1 = textwrap.dedent("""
```
Comment (lasote) [inline conans/test/functional/generators/deploy_test.py]: I would ask all of you to adopt a bit the TurboTestClient and the GenConanfile. You can pass to the latest files you want to package and also the dependencies. The tests are much shorter.

Comment (danimtb) [reply conans/test/functional/generators/deploy_test.py]: ok, let me try

**Comment by danimtb** [comment]
> but it means that generators are being fed with that virtual conanfile instead of name2 conanfile itself, isn't it?

@jgsogo 
Yes, but that is the current behavior in Conan. Packages installed directly by reference are considered the "second" node in the graph and the "virtual consumer" node is created

**Pull request merged**